core/kernel/interrupt.c

```diff
@@ -143,3 +143,125 @@ void __weak __noreturn interrupt_main_handler(void)
 {
 	panic("Secure interrupt handler not defined");
 }
+
+/*
+ * Interrupt controller chip support
+ */
+void interrupt_call_handlers(struct itr_chip *chip, size_t itr_num)
+{
+	struct itr_handler *h = NULL;
+	bool was_handled = false;
+
+	assert(chip);
+
+	SLIST_FOREACH(h, &chip->handlers, link) {
+		if (h->it == itr_num) {
+			if (h->handler(h) == ITRR_HANDLED)
+				was_handled = true;
+			else if (!(h->flags & ITRF_SHARED))
+				break;
+		}
+	}
+
+	if (!was_handled) {
+		EMSG("Disable unhandled interrupt %s:%zu", chip->name, itr_num);
+		interrupt_disable(chip, itr_num);
+	}
+}
+
+TEE_Result interrupt_configure(struct itr_chip *chip, size_t itr_num,
+			       uint32_t type, uint32_t prio)
+{
+	chip->ops->add(chip, itr_num, type, prio);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result interrupt_add_handler(struct itr_handler *hdl)
+{
+	struct itr_handler *h = NULL;
+
+	assert(hdl && hdl->chip->ops);
+
+	SLIST_FOREACH(h, &hdl->chip->handlers, link)
+		if (h->it == hdl->it &&
+		    (!(hdl->flags & ITRF_SHARED) || !(h->flags & ITRF_SHARED)))
+			return TEE_ERROR_GENERIC;
```

Comment:
How is the caller expected to recover from this?

Reply:
I think it is a programming error to mix shared and non-shared interrupt handlers on a same interrupt. At least caller will know something is bad. Existing API function `itr_add_type_prio()` did such verification and panicked in debug mode. I think it is better to simply report the error and let caller panic at its will.

Reply:
Fair enough, please add a debug print.